src/rgw/rgw_op.cc

```diff
@@ -5234,6 +5234,56 @@ int RGWCopyObj::verify_permission(optional_yield y)
 					      RGW_PERM_READ)) { 
 	  return -EACCES;
 	}
+
+  // Check for tagging permissions
+  bool is_put_tagging_permission = false, is_get_tagging_permission = false, without_tagging_directive = false;
+  bool is_tagging_directive_copy = false , is_tagging_directive_replace = false;
+  auto tagging_drctv = s->info.env->get("HTTP_X_AMZ_TAGGING_DIRECTIVE");
+  if (tagging_drctv) {
+    if (strcasecmp(tagging_drctv, "COPY") == 0)
+      is_tagging_directive_copy = true;
+    else if (strcasecmp(tagging_drctv, "REPLACE") == 0)
+      is_tagging_directive_replace = true;
+  } else {
+      without_tagging_directive = true;
+    }
+
+  auto put_tagging_permission = eval_identity_or_session_policies(this,s->iam_user_policies,
```

Comment:
if ( Effect::Allow == eval_identity_or_session_policies(this,s->iam_user_policies, s->env, rgw::IAM::s3PutObjectTagging, obj_arn )
\     tagging_permission |= PUT_TAGGING_PERM

Reply:
Done changes as suggested.

Reply:
the application of an enum class as a bitmap iirc will now fail, you'll need something like set/unset methods that wrap a cast to the underlying type

Reply:
@mattbenjamin we only read these enum values as types, set in a separate var and that is used as a bitmap,  why set/unset methods needed for this enum class?